### PR TITLE
Support google-cloud-spanner v3 and fixes broken unit tests

### LINKED ISSUE #20529: SpannerIO tests don't actually assert anything.
These all have the format

```

p = Pipeline()
...
p.run()
assert_that(...)

```

Imported from Jira [BEAM-10569](https://issues.apache.org/jira/browse/BEAM-10569). Original Jira may contain additional context.
Reported by: robertwb.

### LINKED ISSUE #21198: Google Spanner dependency is too old; cannot deal with JSON type code
Apache Beam is using google-cloud-spanner version 1.19.1, while current version of google-cloud-spanner is 3.11.1.

I was attempting to use ReadFromSpanner when I ran into  a ValueError - it was getting a type code 11 from Spanner and couldn't parse it. 

Here is (some) relevant code from v3.11.1 of Spanner: 
```

class TypeCode(proto.Enum):    
  r"""``TypeCode`` is used as part of [Type][google.spanner.v1.Type]
to
  indicate the type of a Cloud Spanner value.    

  Each legal value of a type can be encoded
to or decoded from a JSON   
  value, using the encodings described below. All Cloud Spanner values
   
  can be ``null``, regardless of type; ``null``\ s are always encoded    
  as a JSON ``null``.

  """
  TYPE_CODE_UNSPECIFIED = 0    
  BOOL = 1    
  INT64 = 2    
  FLOAT64 = 3    
  TIMESTAMP
= 4    
  DATE = 5    
  STRING = 6    
  BYTES = 7    
  ARRAY = 8   
  STRUCT = 9    
  NUMERIC
= 10    
  JSON = 11
```

and the same from v1.19.1:
```

_sym_db.RegisterEnumDescriptor(_TYPECODE)
TypeCode = enum_type_wrapper.EnumTypeWrapper(_TYPECODE)
TYPE_CODE_UNSPECIFIED
= 0
BOOL = 1
INT64 = 2
FLOAT64 = 3T
IMESTAMP = 4
DATE = 5
STRING = 6
BYTES = 7
ARRAY = 8
STRUCT
= 9
NUMERIC = 10 
```

There is no support for type code 11, aka JSON in Apache Beam right now.

Imported from Jira [BEAM-13270](https://issues.apache.org/jira/browse/BEAM-13270). Original Jira may contain additional context.
Reported by: LauraFP.

### PR DESCRIPTION
**Pull request by Abacn**
The native python spannerio was proposed to be deprecated by Q4 but not happens. Still need some maintenance.

Fixes #20529
Fixes #21198

* Fix spannerio read unit test actually not assert anything

* Loose upperbound of spanner client library

**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
`apache_beam/runners/worker/sdk_worker_test.py::SdkWorkerTest::test_harness_monitoring_infos_and_metadata`
will fail if runs after
`apache_beam/io/gcp/experimental/spannerio_test.py::SpannerReadTest::test_read_with_index`
or   
`apache_beam/io/gcp/experimental/spannerio_test.py::SpannerReadTest::test_read_with_table_batch`

Looks like not appropriate teardown/setup for tests

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Performance test passed: https://ci-beam.apache.org/job/beam_PerformanceTests_SpannerIO_Read_2GB_Python_PR/8/
postcommit tests passed (albeit test suite has been timed out on master): https://ci-beam.apache.org/job/beam_PostCommit_Python37_PR/441/testReport/apache_beam.io.gcp.experimental.spannerio_read_it_test/
https://ci-beam.apache.org/job/beam_PostCommit_Python37_PR/441/testReport/apache_beam.io.gcp.experimental.spannerio_write_it_test/
ready for review

R: @johnjcasey 
R: @tvalentyn

**Review by johnjcasey — APPROVED** [review]
LGTM

File: sdks/python/setup.py [hunk 1]
```diff
@@ -289,7 +289,8 @@ def get_portability_package_data():
             'google-cloud-bigquery-storage>=2.6.3,<2.14',
             'google-cloud-core>=0.28.1,<3',
             'google-cloud-bigtable>=0.31.1,<2',
-            'google-cloud-spanner>=1.13.0,<2',
+            # google-cloud-spanner 2.x causes dependency parsing backoff
+            'google-cloud-spanner>=1.13.0,!=2,<=3.21.0',
```
Comment (tvalentyn) [inline sdks/python/setup.py:293]: 1. We shouldn't use <= in the upper bound, it precludes us from asking someone to issue a patch release that could still be picked up by the constraints.
So we can use `<3.22.0` or `<4`. 

2. I think this is server-side dependency only, so using `<4` should be safe as we do with other GCP dependencies at this time. We'll decide soon whether we want to cap upper bounds for all IO.

3. Can you please run PostCommit integration tests with `google-cloud-spanner>=3.0.0` requirement if you haven't ? Unless we specify that requirement, postcommit tests will not pick up the new dependency.

**Comment by tvalentyn** [comment]
cc: @BjornPrime

File: sdks/python/setup.py (hunk 1 above)
Comment (Abacn) [reply sdks/python/setup.py:293]: (1,2) Yes we should set 'google-cloud-spanner<4' as other cloud dependencies

(3) I ran this when debugging (got some error and fixed it). From [logging](https://ci-beam.apache.org/job/beam_PostCommit_Python37_PR/441/consoleFull) it shows version 3.21.0 is picked by pypi:
```
15:20:37 Collecting google-cloud-spanner!=2,<=3.21.0,>=1.13.0
15:20:37   Using cached google_cloud_spanner-3.21.0-py2.py3-none-any.whl (291 kB)
```
also
```
installed: apache-beam @ file:///app/sdks/python/target/.tox/.tmp/package/1/apache-beam-2.43.0.dev0.zip,cachetools==4.2.4,...,google-cloud-recommendations-ai==0.7.1,google-cloud-spanner==3.21.0,...
```
jenkins already cached v3.21.0 and is using it.

Comment (Abacn) [reply sdks/python/setup.py]: Thanks, change applied.

**Comment by Abacn** [comment]
Run Python 3.7 PostCommit

**Comment by Abacn** [comment]
Run SpannerIO Read 2GB Performance Test Python

**Comment by Abacn** [comment]
Integration test and performance test passed with cloud-spanner v3.22.0. Could revert back the last commit

**Pull request merged**